src/main/kotlin/org/rust/ide/structure/RsVisibilitySorter.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.structure
+
+import com.intellij.ide.util.treeView.smartTree.ActionPresentation
+import com.intellij.ide.util.treeView.smartTree.ActionPresentationData
+import com.intellij.ide.util.treeView.smartTree.Sorter
+import org.rust.RsBundle
+import org.rust.ide.icons.RsIcons
+import org.rust.lang.core.psi.ext.RsVisibility
+import org.rust.lang.core.psi.ext.RsVisible
+
+class RsVisibilitySorter : Sorter {
+    private enum class Order {
+        Public,
+        Restricted,
+        Private,
+        Unknown;
+
+        companion object {
+            fun fromVisibility(vis: RsVisibility): Order {
+                return when (vis) {
+                    RsVisibility.Public -> Public
+                    RsVisibility.Private -> Private
+                    is RsVisibility.Restricted -> Restricted
+                }
+            }
+        }
+    }
+
+    private fun getOrdering(x: Any?): Order {
+        val psi = (x as? RsStructureViewElement)?.value
+        val visibility = (psi as? RsVisible)?.visibility
+        return if (visibility != null) Order.fromVisibility(visibility) else Order.Unknown
```

Comment:
What do you think about considering `#[macro_export]` macros as public?

```suggestion
        return when (val psi = (x as? RsStructureViewElement)?.value) {
            is RsVisible -> Order.fromVisibility(psi.visibility)
            is RsMacro -> if (psi.hasMacroExport && !psi.queryAttributes.isDocHidden) Order.Public else Order.Private
            else -> Order.Unknown
        }
```

Reply:
It may be a good idea, but AFAIK currently macros are treated as items with no visibility in the IDE. If we decide to implement it, we'd need to carefully check that the visibility was taken into account in all places.

There are also a couple of edge cases which are currently treated poorly. E.g. a macro `foo!` with no `#[macro_export]` located in `lib::macros` will be treated as private to `lib::macros`. However, we can add `pub(crate) self::foo;` to `lib::macros`, and then `foo` will have Restricted rather than Private visibility (but we cannot add `pub self::foo;` without `#[macro_export]`).

In other words, we would need to properly handle the visibility of macros, and not restrict to a local hack in the Structure View.

Reply:
I would also like to add some sort of visibility to Impl items. I'm not certain yet whether it should be projectwide visibility like for macros, or a local hack. The idea is that, while technically an Impl is visible from anywhere, in practice you can only use it if all involved types and traits are visible for you (i.e. you can see the implementing type, all its generic arguments, the trait with its arguments, and all trait bounds on parameters).

Reply:
A macro without `#[macro_export]` feels like `pub(crate)` with weird name resolution rules (it can't be accessed outside of the module without `#[macro_use]` or re-export in that module).
I suppose it's better to distinguish public macros from private/restricted ones than just show them at the end of the list.